content/backend/graphql-node-typescript/3-graphql-server.md

```diff
@@ -0,0 +1,242 @@
+
+### The complete GraphQL execution pipeline
+
+A complete GraphQL flow is consists of the following steps:
+
+1. Schema building - first, make sure that you have a GraphQL schema ready-to-use (built as executable schema). 
+1. `parse` - takes the GraphQL operation and parse it into a `DocuemntNode`.
+1. `validate` - uses the parsed GraphQL operation with the GraphQL schema, and validate it to make sure it matches the schema.
+1. Variables - GraphQL operations allow you to pass variables, which are just JSON object. 
+1. Context building - GraphQL allow you to pass a custom object, called `context`, and this object will be available for your resolvers. 
+1. `execute` - takes the operation and runs the resovlers based on the selection-set.
+
+With `graphql-helix`, this flow becomes simpler because you can just use `processRequest`.
+
+### Adding GraphiQL
+
+[GraphiQL](https://github.com/graphql/graphiql), a powerful "GraphQL IDE" that runs in your browser, as part of your server, and lets you explore the capabilities of your API in an interactive manner. You can also use it to experiment and test your GraphQL API.
+
+![GraphiQL](https://i.imgur.com/x0WbzF8.png)
+
+`graphql-helix` comes with a ready-to-use GraphiQL configuration that works with any project.
+
+<Instruction>
+
+To add GraphiQL to your server, create a new `GET /graphql` request handler, and just import and use in in a route:
+
+```typescript{3,9-17}(path="hackernews-node-ts/src/index.ts")
+import 'graphql-import-node';
+import fastify from "fastify";
+import { getGraphQLParameters, processRequest, Request, renderGraphiQL, shouldRenderGraphiQL } from "graphql-helix";
+import { schema } from "./schema";
+
+async function main() {
+  const server = fastify();
+
+  server.route({
+    method: ["POST", "GET"],
+    url: "/graphql",
+    handler: async (req, reply) => {
+      const request: Request = {
+        headers: req.headers,
+        method: req.method,
+        query: req.query,
+        body: req.body,
+      };
+
+      if (shouldRenderGraphiQL(request)) {
+        reply.header("Content-Type", "text/html");
+        reply.send(
+          renderGraphiQL({
+            endpoint: "/graphql",
+          })
+        );
+
+        return;
+      }
+
+      const { operationName, query, variables } = getGraphQLParameters(request);
+
+      const result = await processRequest({
+        request,
+        schema,
+        operationName,
+        query,
+        variables,
+      });
+
+      if (result.type === "RESPONSE") {
+        reply.send(result.payload);
+      } else {
+        reply.send({ error: "Stream not supported at the moment" });
+      }
+    }
+  });
+
+  server.listen(3000, "0.0.0.0");
+}
+
+main();
+```
+
+</Instruction>
+
+Now, run the server, and try to access `http://localhost:3000/graphql` in your browser. You should see GraphiQL UI, and from there you can explore the schema, the documentation, and even try to run queries! 
```

Comment:
You should provide a screenshot of what the user should expect to see. Furthermore, you should explain some of the "basics" of the graphiql UI so the user understands what they are looking at.